### PR TITLE
Feature/fix duplicate characters

### PR DESCRIPTION
**Pull request by Acidity**
Resolves Issues #31, #33, #34, and #46.

### CODE REVIEW

**Comment by Acidity** [comment]
I've tested this personally, but would appreciate someone else testing to confirm functionality.

File: brave/core/key/controller.py [hunk 1]
```diff
@@ -89,6 +89,10 @@ def post(self, **kw):
                 
         except ValueError:
             return 'json:', dict(success=False, message=_("Key ID must be a number."), field='key')
+            
+        #Prevent the user from adding a key that has already been added.
+        if EVECredential.objects(key=data.key):
```
Comment (jboning) [inline brave/core/key/controller.py]: This is racy--if two accounts add the same key at the same time, they can both pass this check and then both add objects. Better to make key unique and then check for failure at save time.

File: brave/core/key/template/list.html [hunk 2]
```diff
@@ -89,8 +89,8 @@
                                         '</button>' +
                                     '</td>' +
                                 '</tr>'
-			).appendTo('table.table tbody');
-			console.log("ADDED TO TBODY");
+                        ).appendTo('table.table tbody');
+                        console.log("ADDED TO TBODY");
```
Comment (jboning) [inline brave/core/key/template/list.html]: probably OK to clean this up

File: brave/core/key/template/list.html [hunk 3]
```diff
@@ -140,6 +140,11 @@
     <tr data-id="${record.id}">
         <td class="key-id"><i class="fa fa-fw fa-${'check' if record.verified else 'times'}"></i> ${record.key}</td>
         <td class="key-code"><span class="ellipsis">${record.code}</span></td>
+        % if record.violation == "Character":
+        <td class="security-violation"><font color="red">Character Security Violation.</font></td>
```
Comment (jboning) [inline brave/core/key/template/list.html]: I'd prefer a more descriptive error message

File: brave/core/key/template/list.html [hunk 4]
```diff
@@ -140,6 +140,11 @@
     <tr data-id="${record.id}">
```
Comment (jboning) [inline brave/core/key/template/list.html:140]: is row() actually used? very similar stuff seems to be open-coded below.

**Comment by jboning** [comment]
The row that gets added to the table ajaxily does not include the violation cell

**Comment by jboning** [comment]
There's a complete lack of feedback when submitting an API key that's already on a different account. #145 :/

File: brave/core/key/model.py [hunk 5]
```diff
@@ -177,12 +185,18 @@ def pull(self):
...
         
+        allCharsOK = True
+
         for char in result.characters.row:
             if 'corporationName' not in char:
                 log.error("corporationName missing for key %d", self.key)
                 continue
             
-            implementation(char)
+            if implementation(char) == (None, None, None):
```
Comment (jboning) [inline brave/core/key/model.py]: This isn't the only place that calls pull_minimal -- grep for other call sites and fix those too.

Comment (jboning) [inline brave/core/key/model.py]: On second though, hold off on fixing these. I'm about to tear up and refactor these functions in a big way, so don't worry about it.

**Comment by Acidity** [comment]
Accidentally pushed this too early. Still has issues.

**Comment by Acidity** [comment]
Still no feedback when attempting to add a key that's taken, but I'd rather leave that to another PR.

**Comment by Acidity** [comment]
Replacing with new pull request.

**Pull request closed without merging**